pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/MultiStageQueryStats.java

```diff
@@ -0,0 +1,668 @@
+   * current stage.
+   */
+  public void mergeUpstream(MultiStageQueryStats otherStats) {
+    Preconditions.checkArgument(_currentStageId <= otherStats._currentStageId,
+        "Cannot merge stats from early stage %s into stats of later stage %s",
+        otherStats._currentStageId, _currentStageId);
+
+    growUpToStage(otherStats.getMaxStageId());
+
+    int currentDiff = otherStats._currentStageId - _currentStageId;
+    if (currentDiff > 0) {
+      StageStats.Closed close = otherStats._currentStats.close();
+      int selfIdx = currentDiff - 1;
+      StageStats.Closed myStats = _closedStats.get(selfIdx);
+      if (myStats == null) {
+        _closedStats.set(selfIdx, close);
+      } else {
+        myStats.merge(close);
+      }
+    }
+
+    for (int i = 0; i < otherStats._closedStats.size(); i++) {
+      StageStats.Closed otherStatsForStage = otherStats._closedStats.get(i);
+      if (otherStatsForStage == null) {
+        continue;
+      }
+      int selfIdx = i + currentDiff;
+      StageStats.Closed myStats = _closedStats.get(selfIdx);
+      try {
+        if (myStats == null) {
+          _closedStats.set(selfIdx, otherStatsForStage);
+          assert getUpstreamStageStats(i + otherStats._currentStageId + 1) == otherStatsForStage;
+        } else {
+          myStats.merge(otherStatsForStage);
+        }
+      } catch (IllegalArgumentException | IllegalStateException ex) {
+        LOGGER.warn("Error merging stats on stage " + i + ". Ignoring the new stats", ex);
+      }
+    }
+  }
+
+  public void mergeUpstream(List<ByteBuffer> otherStats) {
+    for (int i = 0; i <= _currentStageId && i < otherStats.size(); i++) {
+      if (otherStats.get(i) != null) {
+        throw new IllegalArgumentException("Cannot merge stats from early stage " + i + " into stats of "
+            + "later stage " + _currentStageId);
+      }
+    }
+    growUpToStage(otherStats.size() - 1);
+
+    for (int i = _currentStageId + 1; i < otherStats.size(); i++) {
+      ByteBuffer otherBuf = otherStats.get(i);
+      if (otherBuf != null) {
+        StageStats.Closed myStats = getUpstreamStageStats(i);
+        try (InputStream is = new ByteBufferInputStream(Collections.singletonList(otherBuf));
+            DataInputStream dis = new DataInputStream(is)) {
+          if (myStats == null) {
+            StageStats.Closed deserialized = StageStats.Closed.deserialize(dis);
+            _closedStats.set(i - _currentStageId - 1, deserialized);
+            assert getUpstreamStageStats(i) == deserialized;
+          } else {
+            myStats.merge(dis);
+          }
+        } catch (IOException ex) {
+          boolean assertOn = false;
+          // *assigns* true if assertions are on.
+          //CHECKSTYLE:OFF
+          assert assertOn = true;
+          if (assertOn) {
+            throw new UncheckedIOException("Error deserializing stats on stage " + i, ex);
+          }
+          //CHECKSTYLE:ON
```

Comment:
This is the _standard_ way to detect whether asserts are enabled or not. Maybe we should create a static class that does the same thing, specially given checkstyle doesn't like this code style

Reply:
Is the intention to fail it during testing?
Having different behavior with/without assert on seems weird to me though

Reply:
I feel it is safer to have consistent behavior in test and production. In the test we can check for the stats value to verify it is correctly handled.
I don't have strong opinion, so either way is fine